include/rear-panel/core/tasks/host_comms_task.hpp

```diff
@@ -0,0 +1,134 @@
+
+#include "FreeRTOS.h"
+#include "common/core/message_queue.hpp"
+#include "rear-panel/core/double_buffer.hpp"
+#include "rear-panel/core/messages.hpp"
+
+namespace host_comms_task {
+
+using TaskMessage =
+    std::variant<std::monostate, messages::IncomingMessageFromHost>;
+
+template <class ResposneQueue>
+class HostCommMessageHandler {
+  public:
+    explicit HostCommMessageHandler(ResposneQueue &resp_queue)
+        : resp_queue{resp_queue} {}
+    HostCommMessageHandler(const HostCommMessageHandler &) = delete;
+    HostCommMessageHandler(const HostCommMessageHandler &&) = delete;
+    auto operator=(const HostCommMessageHandler &)
+        -> HostCommMessageHandler & = delete;
+    auto operator=(const HostCommMessageHandler &&)
+        -> HostCommMessageHandler && = delete;
+    ~HostCommMessageHandler() = default;
+
+    template <typename InputIt, typename InputLimit>
+    requires std::forward_iterator<InputIt> &&
+        std::sized_sentinel_for<InputLimit, InputIt>
+    auto handle_message(TaskMessage &m, InputIt tx_into, InputLimit tx_limit)
+        -> InputIt {
+        // we need a this-capturing lambda to pass on the call to our set of
+        // member function overloads because otherwise we would need a pointer
+        // to member function, and you can't really do that with variant visit.
+        // we also need to current the transmit buffer details in.
+        auto visit_helper = [this, tx_into, tx_limit](auto &m) -> InputIt {
+            return this->visit_message(m, tx_into, tx_limit);
+        };
+
+        // now, calling visit on the visit helper will pass through the calls to
+        // our message handlers, and will pass through whatever the messages
+        // return (aka how much data they wrote, if any) to the caller.
+        return std::visit(visit_helper, m);
+    }
+
+  private:
+    template <typename InputIt, typename InputLimit>
+    requires std::forward_iterator<InputIt> &&
+        std::sized_sentinel_for<InputLimit, InputIt>
+    auto visit_message(std::monostate &m, InputIt tx_into, InputLimit tx_limit)
+        -> InputIt {
+        static_cast<void>(m);
+        static_cast<void>(tx_into);
+        static_cast<void>(tx_limit);
+        return tx_into;
+    }
+
+    // for this very basic first pass we just want to echo back any messages we
+    // receive from the host
+    template <typename InputIt, typename InputLimit>
+    requires std::forward_iterator<InputIt> &&
+        std::sized_sentinel_for<InputLimit, InputIt>
+    auto visit_message(messages::IncomingMessageFromHost &msg, InputIt tx_into,
+                       InputLimit tx_limit) -> InputIt {
+        return std::copy(msg.buffer,
+                         std::min(msg.limit, msg.buffer + (tx_into - tx_limit)),
+                         tx_into);
+    }
+
+    ResposneQueue &resp_queue;
+};
+
+/**
+ * The task type.
+ */
+template <template <class> class QueueImpl>
+requires MessageQueue<QueueImpl<TaskMessage>, TaskMessage>
```

Comment:
we might need to rework this into the ot3-firmware style with a message handler idk